azure/scope/identity.go

```diff
@@ -361,3 +371,112 @@ func IsClusterNamespaceAllowed(ctx context.Context, k8sClient client.Client, all
 
 	return false
 }
+
+func reconcileASOSecret(ctx context.Context, azureClusterIdentity *infrav1.AzureClusterIdentity, kubeClient client.Client, subscriptionID string) error {
+	asoSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      identity.GetASOSecretName(azureClusterIdentity.Name, subscriptionID),
+			Namespace: azureClusterIdentity.Namespace,
+			Labels: map[string]string{
+				infrav1.ASOSecretLabel: azureClusterIdentity.Name,
+			},
+		},
+		Data: map[string][]byte{
+			"AZURE_SUBSCRIPTION_ID": []byte(subscriptionID),
+			"AZURE_TENANT_ID":       []byte(azureClusterIdentity.Spec.TenantID),
+			"AZURE_CLIENT_ID":       []byte(azureClusterIdentity.Spec.ClientID)},
+	}
+
+	apiVersion, kind := infrav1.GroupVersion.WithKind(azureClusterIdentity.Kind).ToAPIVersionAndKind()
+	owner := metav1.OwnerReference{
+		APIVersion: apiVersion,
+		Kind:       kind,
+		Name:       azureClusterIdentity.GetName(),
+		UID:        azureClusterIdentity.GetUID(),
+	}
+
+	// Fetch identity secret, if it exists
+	key := types.NamespacedName{
+		Namespace: azureClusterIdentity.Spec.ClientSecret.Namespace,
+		Name:      azureClusterIdentity.Spec.ClientSecret.Name,
+	}
+	identitySecret := &corev1.Secret{}
+	err := kubeClient.Get(ctx, key, identitySecret)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return errors.Errorf("failed to fetch identity secret %s as defined in AzureClusterIdentity %q/%q: %v", identitySecret.Name, azureClusterIdentity.Namespace, azureClusterIdentity.Name, err)
+	}
+
+	// TODO add support for workload identity
+	switch azureClusterIdentity.Spec.Type {
+	case infrav1.ServicePrincipal, infrav1.ManualServicePrincipal:
+		asoSecret.Data["AZURE_CLIENT_SECRET"] = identitySecret.Data["clientSecret"]
+	case infrav1.ServicePrincipalCertificate:
+		asoSecret.Data["AZURE_CLIENT_CERTIFICATE"] = identitySecret.Data["certificate"]
+		asoSecret.Data["AZURE_CLIENT_CERTIFICATE_PASSWORD"] = identitySecret.Data["password"]
+	case infrav1.UserAssignedMSI:
+		asoSecret.Data["AZURE_IDENTITY_RESOURCE_ID"] = []byte(azureClusterIdentity.Spec.ResourceID)
+	}
+
+	// Fetch previous secret, if it exists
+	key = types.NamespacedName{
+		Namespace: asoSecret.Namespace,
+		Name:      asoSecret.Name,
+	}
+	oldSecret := &corev1.Secret{}
+	err = kubeClient.Get(ctx, key, oldSecret)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return errors.Wrap(err, "failed to fetch existing ASO secret")
+	}
+
+	// Create if it wasn't found
+	if apierrors.IsNotFound(err) {
+		if err := kubeClient.Create(ctx, asoSecret); err != nil && !apierrors.IsAlreadyExists(err) {
+			return errors.Wrap(err, "failed to create ASO secret")
+		}
+		return nil
+	}
+
+	// Otherwise, check ownership and data freshness. Update as necessary
+	hasOwner := false
+	for _, ownerRef := range oldSecret.OwnerReferences {
+		if referSameObject(ownerRef, owner) {
```

Comment:
Is there a case where only checking the UID wouldn't work here?

Reply:
Maybe I'm misunderstanding the question, but we need to make sure the old secret has all the necessary stuffs in place(owner, data ,etc), so not sure how only checking the UID might work.

Reply:
I mean using only the UID to see if `ownerRef` and `owner` are equivalent instead of the combination of API version, kind, and name like `referSameObject` does.